### PR TITLE
Feature/improve alignment equations

### PR DESCRIPTION
**Pull request by bloesch**
The main changes are in the alignment equations. This comes with a couple of new variables which can be set in the info file. All in all performance should remain similar. The outlier handling is more strict.

### CODE REVIEW

File: cfg/rovio.info [hunk 1]
```diff
@@ -96,8 +97,8 @@ ImgUpdate
     statLocalQualityRange 10;									Number of frames for local quality evaluation
     statLocalVisibilityRange 100;								Number of frames for local visibility evaluation
     statMinGlobalQualityRange 100;								Minimum number of frames for obtaining maximal global quality
-    trackingUpperBound 0.9;										Threshold for local quality for min overall global quality
-    trackingLowerBound 0.1;										Threshold for local quality for max overall global quality
+    trackingUpperBound 0.999;										Threshold for local quality for min overall global quality
+    trackingLowerBound 0.999;										Threshold for local quality for max overall global quality
```
Comment (burrimi) [inline cfg/rovio.info]: intended?

Comment (bloesch) [inline cfg/rovio.info]: jop

File: include/rovio/CoordinateTransform/LandmarkOutput.hpp [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    mpMultiCamera_ = mpMultiCamera;
+    ID_ = -1;
+  };
+  virtual ~LandmarkOutputImuCT(){};
+  void setFeatureID(int ID){
+    ID_ = ID;
+  }
+  void evalTransform(mtOutput& output, const mtInput& input) const{
+    input.updateMultiCameraExtrinsics(mpMultiCamera_);
+    // MrMP = MrMC + qCM^T(d_in*nor_in)
```
Comment (burrimi) [inline include/rovio/CoordinateTransform/LandmarkOutput.hpp]: match comment with implementation
`CrCP = ...`

File: include/rovio/CoordinateTransform/LandmarkOutput.hpp [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    const V3D CrCP = input.dep(ID_).getDistance()*input.CfP(ID_).get_nor().getVec();
+    output.template get<mtOutput::_lmk>() = mpMultiCamera_->BrBC_[input.CfP(ID_).camID_] + mpMultiCamera_->qCB_[input.CfP(ID_).camID_].inverseRotate(CrCP);
+  }
+  void jacTransform(MXD& J, const mtInput& input) const{
+    J.setZero();
+    input.updateMultiCameraExtrinsics(mpMultiCamera_);
+    const V3D CrCP = input.dep(ID_).getDistance()*input.CfP(ID_).get_nor().getVec();
+    const Eigen::Matrix<double,3,2> J_CrCP_nor = input.dep(ID_).getDistance()*input.CfP(ID_).get_nor().getM();
+    const Eigen::Matrix<double,3,1> J_CrCP_d = input.CfP(ID_).get_nor().getVec()*input.dep(ID_).getDistanceDerivative();
+    const M3D mMC = MPD(mpMultiCamera_->qCB_[input.CfP(ID_).camID_].inverted()).matrix();
```
Comment (burrimi) [inline include/rovio/CoordinateTransform/LandmarkOutput.hpp]: consistent notation `const M3D mBC = ...`

File: include/rovio/FeatureManager.hpp [hunk 4]
```diff
@@ -387,6 +390,102 @@ class FeatureSetManager{
...
   }
+
+  static bool isFirstMPBetter(const std::tuple<const FeatureCoordinates*,MultilevelPatch<nLevels,patchSize>,int>& mp1, const std::tuple<const FeatureCoordinates*,MultilevelPatch<nLevels,patchSize>,int>&mp2){
+    return std::get<1>(mp1).s_ > std::get<1>(mp2).s_;
+  }
+
+  std::unordered_set<unsigned int> addBestCandidatesNew(const std::vector<FeatureCoordinates>& candidates, const ImagePyramid<nLevels>& pyr, const int camID, const double initTime,
+                                                       const int l1, const int l2, const int maxN, const int nDetectionBuckets, const double scoreDetectionExponent,
+                                                       const double penaltyDistance, const double zeroDistancePenalty, const bool requireMax, const float minScore){
+      std::unordered_set<unsigned int> newSet;
```
Comment (burrimi) [inline include/rovio/FeatureManager.hpp]: rename `bestCandidates`.

File: include/rovio/FeatureManager.hpp [hunk 5]
```diff
@@ -387,6 +390,102 @@ class FeatureSetManager{
 
     return newSet;
   }
+
+  static bool isFirstMPBetter(const std::tuple<const FeatureCoordinates*,MultilevelPatch<nLevels,patchSize>,int>& mp1, const std::tuple<const FeatureCoordinates*,MultilevelPatch<nLevels,patchSize>,int>&mp2){
+    return std::get<1>(mp1).s_ > std::get<1>(mp2).s_;
+  }
+
+  std::unordered_set<unsigned int> addBestCandidatesNew(const std::vector<FeatureCoordinates>& candidates, const ImagePyramid<nLevels>& pyr, const int camID, const double initTime,
+                                                       const int l1, const int l2, const int maxN, const int nDetectionBuckets, const double scoreDetectionExponent,
```
Comment (burrimi) [inline include/rovio/FeatureManager.hpp]: is maxN the same as nMax?

File: include/rovio/MultilevelPatchAlignment.hpp [hunk 6]
```diff
@@ -129,12 +202,93 @@ class MultilevelPatchAlignment {
...
+              *it_dy_error = reg_a*(*it_dy_error - reg_a_dy*(*it_patch) - reg_b_dy);
+            }
+            b((numLevel-1)*patch_size*patch_size+y*patch_size+x,0) = *it_error;
+            A((numLevel-1)*patch_size*patch_size+y*patch_size+x,0) = *it_dx_error;
+            A((numLevel-1)*patch_size*patch_size+y*patch_size+x,1) = *it_dy_error;
+
+            if(huberNormThreshold_ > 0.0){ // TODO: investigate why 1.0 leads to non-deterministic behavior
+              const float b_abs = std::fabs(*it_error);
+              if(b_abs > huberNormThreshold_){
+                b((numLevel-1)*patch_size*patch_size+y*patch_size+x,0) = std::sqrt(huberNormThreshold_*(2.0*b_abs - huberNormThreshold_));
```
Comment (burrimi) [inline include/rovio/MultilevelPatchAlignment.hpp]: I think you need to add the sign
`b((numLevel-1)*patch_size*patch_size+y*patch_size+x,0) = std::sqrt(huberNormThreshold_*(2.0*b_abs - huberNormThreshold_))*copysign(1.0, *it_error);`

**Comment by burrimi** [comment]
scolgtm.

File: include/rovio/MultilevelPatchAlignment.hpp (hunk 6 above)
Comment (bloesch) [inline include/rovio/MultilevelPatchAlignment.hpp]: should be alright, changed in the subsequent PR

**Pull request merged**